### ISSUE TITLE
Is it possible to set marked extension directly?

### ISSUE THREAD

**Issue by wlnirvana** [hakimel/reveal.js#3202]
# TL;DR

 `Reveal.initialize({markdown:{render: myRenderer}})`  does not extend the default renderer, but completely replace it with the argument passed.

# Details

I'm trying to integrate mermaidjs into reveal by setting up an extended renderer to marked. However, the following code does not work as expected:

````html
<!doctype html>
<html>

<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no">

    <title>Test</title>

    <link href="https://lf3-cdn-tos.bytecdntp.com/cdn/expire-1-M/reveal.js/4.1.0/reset.min.css" type="text/css" rel="stylesheet" />
    <link href="https://lf26-cdn-tos.bytecdntp.com/cdn/expire-1-M/reveal.js/4.1.0/reveal.css" type="text/css" rel="stylesheet" />
    <link href="https://lf26-cdn-tos.bytecdntp.com/cdn/expire-1-M/reveal.js/4.1.0/theme/black.min.css" type="text/css" rel="stylesheet" />
    <link href="https://lf26-cdn-tos.bytecdntp.com/cdn/expire-1-M/reveal.js/4.1.0/plugin/highlight/monokai.min.css" type="text/css" rel="stylesheet" />
</head>

<body>
    <div class="reveal">
        <div class="slides">
            <section data-markdown data-separator="^\n\n\n">
                <textarea data-template>
# Mermaid Graph

```mermaid
graph TD;
    A-->B;
    A-->C;
    B-->D;
    C-->D;
```
		</textarea>
            </section>
        </div>
    </div>

    <script src="https://lf3-cdn-tos.bytecdntp.com/cdn/expire-1-M/reveal.js/4.1.0/reveal.min.js" type="application/javascript"></script>
    <script src="https://lf26-cdn-tos.bytecdntp.com/cdn/expire-1-M/reveal.js/4.1.0/plugin/notes/notes.min.js" type="application/javascript"></script>
    <script src="https://lf3-cdn-tos.bytecdntp.com/cdn/expire-1-M/reveal.js/4.1.0/plugin/markdown/markdown.min.js" type="application/javascript"></script>
    <script src="https://lf26-cdn-tos.bytecdntp.com/cdn/expire-1-M/reveal.js/4.1.0/plugin/highlight/highlight.min.js" type="application/javascript"></script>
    <script src="https://lf3-cdn-tos.bytecdntp.com/cdn/expire-0-M/mermaid/8.14.0/mermaid.min.js" type="application/javascript"></script>
    <script>
        mermaid.initialize({
            startOnLoad: true
        });
    </script>
    <script>
        const renderer = {
            code(code, language, escaped) {
                if (code.match(/^sequenceDiagram/) || code.match(/^graph/)) {
                    return '<div class="mermaid">' + code + '</div>';
                } else {
                    return '<pre><code>' + code + '</code></pre>';
                }
            }
        };
        Reveal.initialize({
            hash: true,
            transition: 'none',
            slideNumber: true,
            markdown: {
                renderer
            },
            plugins: [RevealMarkdown, RevealHighlight, RevealNotes]
        });
    </script>
</body>

</html>
````

The error message below indicates that `Reveal.initialize(...)` does not internally call `marked.use` to [merge](https://github.com/markedjs/marked/issues/2082#issuecomment-851704919) the renderer, but replaces the default renderer with a brand new object.

```text
Uncaught (in promise) TypeError: t.text is not a function
Please report this to https://github.com/markedjs/marked.
    at as.value (markdown.min.js:1:62962)
    at as.value (markdown.min.js:1:60731)
    at Function.value (markdown.min.js:1:63175)
    at fs (markdown.min.js:1:65090)
    at markdown.min.js:1:70592
    at Array.forEach (<anonymous>)
    at s (markdown.min.js:1:70474)
```

To work around this issue, I have to manually introduce marked script, call `new marked.Renderer();` to construct a default renderer, extend it, and finally "inject" it into reveal.js, like in the following code snippet.

````html
<!doctype html>
<html>

<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no">

    <title>Test</title>

    <link href="https://lf3-cdn-tos.bytecdntp.com/cdn/expire-1-M/reveal.js/4.1.0/reset.min.css" type="text/css" rel="stylesheet" />
    <link href="https://lf26-cdn-tos.bytecdntp.com/cdn/expire-1-M/reveal.js/4.1.0/reveal.css" type="text/css" rel="stylesheet" />
    <link href="https://lf26-cdn-tos.bytecdntp.com/cdn/expire-1-M/reveal.js/4.1.0/theme/black.min.css" type="text/css" rel="stylesheet" />
    <link href="https://lf26-cdn-tos.bytecdntp.com/cdn/expire-1-M/reveal.js/4.1.0/plugin/highlight/monokai.min.css" type="text/css" rel="stylesheet" />
</head>

<body>
    <div class="reveal">
        <div class="slides">
            <section data-markdown data-separator="^\n\n\n">
                <textarea data-template>
# Mermaid Graph

```mermaid
graph TD;
    A-->B;
    A-->C;
    B-->D;
    C-->D;
```
		</textarea>
            </section>
        </div>
    </div>

    <script src="https://lf3-cdn-tos.bytecdntp.com/cdn/expire-1-M/reveal.js/4.1.0/reveal.min.js" type="application/javascript"></script>
    <script src="https://lf26-cdn-tos.bytecdntp.com/cdn/expire-1-M/reveal.js/4.1.0/plugin/notes/notes.min.js" type="application/javascript"></script>
    <script src="https://lf3-cdn-tos.bytecdntp.com/cdn/expire-1-M/reveal.js/4.1.0/plugin/markdown/markdown.min.js" type="application/javascript"></script>
    <script src="https://lf26-cdn-tos.bytecdntp.com/cdn/expire-1-M/reveal.js/4.1.0/plugin/highlight/highlight.min.js" type="application/javascript"></script>
    <script src="https://lf3-cdn-tos.bytecdntp.com/cdn/expire-0-M/mermaid/8.14.0/mermaid.min.js" type="application/javascript"></script>
    <script src="https://lf9-cdn-tos.bytecdntp.com/cdn/expire-0-M/marked/4.0.2/marked.min.js" type="application/javascript"></script>
    <script>
        mermaid.initialize({
            startOnLoad: true
        });
    </script>
    <script>
        const renderer = new marked.Renderer();
        renderer.code = function(code, language, escaped) {
            if (language.match(/^mermaid/)) {
                return '<div class="mermaid">' + code + '</div>';
            } else {
                return '<pre><code>' + code + '</code></pre>';
            }
        };
        Reveal.initialize({
            hash: true,
            transition: 'none',
            slideNumber: true,
            markdown: {
                renderer
            },
            plugins: [RevealMarkdown, RevealHighlight, RevealNotes]
        });
    </script>
</body>

</html>
````

This seems quite awkward, especially considering reveal.js already includes marked. Did I miss anything?

**Comment by jefBinomed** [comment]
After two days, I find an "official" way to integrate marked extensions to revealJS. It's finally easy.

According to the fact that your extension is respecting marked ^4.3.0 version (when I write this comment, it's version 15)

```html
 <script>
// Define your marked Extension (something like that, I don't test it)
const mermaidExt = {
  name: 'mermaidExt',
  level: 'block',                                     // Is this a block-level or inline-level tokenizer?
  start(src) { return src.match(/^mermaid/)?.index; }, // Hint to Marked.js to stop and check for a match
  tokenizer(src, tokens) {
    const match = /^mermaid( (\w+))?/.exec(src);
    if (match) {
      const className = match?.[2];
      const rows = src.split('\n');
      const nextRows = rows.slice(1);
      const closingTokenIndex = nextRows.findIndex((row) =>
        row.match(/^```/),
      );
      const closingToken =
        closingTokenIndex === -1
          ? ''
          : '\n' + nextRows[closingTokenIndex];
      const sliceEnd =
        closingTokenIndex === -1 ? undefined : closingTokenIndex;
      const divContent = nextRows.slice(0, sliceEnd);
      const rawContent = divContent.join('\n');
      const token: mermaidToken = {
        type: 'mermaidExt',
        raw: `${rows[0]}\n${rawContent}${closingToken}`,
        class: className,
        tokens: [],
      };
      this.lexer.blockTokens(rawContent, token.tokens);
      return token;
    }
  },
  renderer(token) {
    return '<div class="mermaid">' + token + '</div>';; // parseInline to turn child tokens into HTML
  }
};

// Initialize Reveal Markdown plugin outside and adding to it your marked extensions !
const revealMarkdownPlugin = RevealMarkdown();
revealMarkdownPlugin.marked.use(mermaidExt);       

// Simply initialize reveal
Reveal.initialize({
  hash: true,
  transition: 'none',
  slideNumber: true,
  plugins: [
    () => {
      id: 'markdown-with-extensions',
      init(reveal) {
        return revealMarkdownPlugin.init(reveal)
      }
    },
    RevealHighlight, 
    RevealNotes
  ]
});

</script>
```
